src/amqp-client/channel.cr

```diff
@@ -286,32 +304,51 @@ class AMQP::Client
       raise ClosedException.new(@closing_frame) if @closing_frame
     end
 
-    @on_confirm = Sync(Hash(UInt64, Proc(Bool, Nil))).new
+    @unconfirmed_publishes = Deque(UInt64).new
+    @unconfirmed_empty = ::Channel(Bool).new
+    @confirm_lock = Mutex.new
     @last_confirm = {0_u64, true}
 
+    @[Deprecated("Use `#wait_for_confirms` instead")]
     def on_confirm(msgid, &blk : Bool -> Nil)
       raise ArgumentError.new "Confirm id must be > 0" unless msgid > 0
       last_confirm, last_confirm_ok = @last_confirm
       if last_confirm >= msgid.to_u64
         blk.call last_confirm_ok
```

Comment:
this assumes that ack's are coming in order, so it's probably a fault implementation. a first step could be to check if the last `ack` was sent with the `multiple` flag, then this holds.

Reply:
Right, we have this bug in the current release. I'll always wait for all confirms to keep it simple.